### ISSUE TITLE
[Quesion] Javascript Framework choice

### ISSUE THREAD

**Issue by aRn0D** [Sylius/Sylius#228]
We could start to speak about the javascript framework choice. What do you think about http://knockoutjs.com/ and the pattern MVVM?

**Comment by aramalipoor** [comment]
+1 for StorefrontBundle in its traditional twig form.
And
+1 for AngularJS + RESTful API for Backend.

I've integrated Angular with a Symfony2 project lately and I found that there are a lot of ideas for abstraction that can ease the process of developing Sylius Backend as well as publishing a Bundle or two in the wild so they can be used in all Angular+Symfony integration projects.

E.g. FormType to [Schema.json](http://json-schema.org/examples.html) converter bundle which creates a Json definition of a Symfony FormType.

IMO we can wire all Symfony's power (Validation, Intl, Form, ACL, etc.) to Angular world. Not only useful for Sylius Backend but other people too. Of course it's a lot of work. But I think it would be awesome.

**Comment by kayue** [comment]
@aramalipoor Schema.json sounds like an interesting solution.

My concern is it won't incorporate with form theme in Symfony.

**Comment by mosch** [comment]
I would choose [React](http://facebook.github.io/react/) and a RESTful API. Active community, lot's of reusable components and really fast and fun.

**Comment by steffenbrem** [comment]
I am building an API around Sylius and currently use [http://emberjs.com/](Ember) for the client side, this is by far the best pick for such a big application as Sylius is. You should take a look at this, it is already very mature!

**Comment by kayue** [comment]
@mosch React isn't template driven, which isn't very a very good fit for Sylius IMO.

I think **ideally** Sylius's JS framework should allow developer to build awesome UI without modifying a lot of HTML (view), just focus on JS and PHP.

Therefor I think AngularJS is the closest one because it is template / attribute driven. (Web Components is even better if no browser drawback..)

**Comment by mosch** [comment]
@kayue you're right. react isn't template driven - but component driven. and it perfectly fits your requirements since stylius could provide a lot of configurable and reusable ready-to-use components. 
The developer can either just use the ready to use components and just pass attributes to them to configure them, or they use some child components for maximum flexibility.

**Comment by RuslanZavacky** [comment]
Hey guys, I am trying to follow how you develop Sylius, and I would like to give you some hints on choosing the right framework for your Job. Consider reading Investigation that was done by Ghost team about choosing a framework for their admin, they've already spent a lot of time and a lot of people from Angular world vs Ember world where involved in the discussions.

**Main thread**
https://github.com/TryGhost/Ghost/issues/2144

**Angular.js Invetigation**
https://github.com/TryGhost/Ghost/issues/2201

**React Investigation**
https://github.com/TryGhost/Ghost/issues/2203

**Ember.js Investigation**
https://github.com/TryGhost/Ghost/issues/2200

To sum up, they've decided to go with Ember.js. The speed, on how Ember is evolving is magnificent. The HTMLBars + Fastboot will land into Ember in 2015 and will make this framework really good for backend as Admin UI and also for Frontend as it will be available to render whole ember app in the backend.

**Comment by kayue** [comment]
Thanks @RuslanZavacky , good reference. Would you recommend Ember if we are not doing a single page app?

BTW, It is also worth a look to [OroCrm's front-end architecture](http://www.orocrm.com/documentation/index/current/book/frontend-architecture), since they are also built with Symfony.

**Comment by RuslanZavacky** [comment]
@kayue My suggestion might be the following:

* Separate repository to hold front-end of admin
* API of admin exists in the Sylius
* Admin frontend as sub-module added to Sylius.

Separation of concerns and separate people can work on it. And for that you have to go with API First Design and this approach will help other people/companies who want to manipulate admin of their e-commerce stores from their internal systems.

And when you work with Frontend as "real" front-end, which is not tied to back-end, you will develop it differently, and can use more tools to speed up development.

About your question "Would you recommend Ember if we are not doing a single page app?" Currently, there is no good solution, to create normal web page app that will be crawlable by Search Engines. Prerender.io can be used (and I am actually using it now to generate HTML representation of ember.js app, so I am kinda using it for not single-page app). But this approach with prerender.io brings different problems, you have to prepare Sitemap and HTML for pages before Google visits your page and then serve it statically, because each render with prerender.io takes a lot of time, sometimes till 7 seconds. Ember's FastBoot solution will solve this problem, but will be released <who-knows-when> in this year and I am really waiting for it.

**Comment by kayue** [comment]
@RuslanZavacky Thanks for introducing Ember's FastBoot, and yes this is the solution I want to use. I guess it probably won't be ready to use until the end of the year.

I was checking Ember's components, one thing I don't like about it is it requires handlebars. I think AngularJS's attribute approach is better in this aspect.

PS: Sylius probably won't be using any JS framework for store front, so SEO isn't one of the concern.

**Comment by RuslanZavacky** [comment]
@kayue very soon handlebars will be vanished from the Ember :) HTMLBars are almost there. There is and performance fiddle to compare some of the implementations: http://jsfiddle.net/Ut2X6/

Problem in Angular from my perspective is approach, that it starts from the HTML, and you have to place Angular on top of HTML.

Ember starts from the side of architecture and separation of concerns. So you have Routes/Controllers/Components that translates into HTML.

Angular plays in to different direction. One more thing is that Angular will release 2nd version, and its very different from the first one. So if you decide to start with Angular, you have to go with 2nd version. As Admin interface is not that you will rewrite often, then its better to choose something, that have more solid support & migration plan.

**Comment by torinaki** [comment]
Thank you @RuslanZavacky for great constructive comments and advise. Unfortunately I do not have any experience with Ember, but I have have few comments about Angular. 

> So if you decide to start with Angular, you have to go with 2nd version

I'm not pretty agree with this, because:
The second version of Angular will not be release in near future. According to last Angular team [planning meeting](http://angularjs.blogspot.com/2014/12/planning-angular-14.html) there are at least 3 versions of Angular 1 should be released. The current state of Angular versions:
1.2.x - the latest stable version
1.3.x - not yet stable
1.4.x - planning(http://www.infoq.com/news/2014/12/angularjs-14)
1.5.x - according to [plan](https://docs.google.com/spreadsheets/d/1ihM8ORf8v38qAbnzJM4TiCqdKeCRz_1VO9JPjSpPWKU/edit#gid=0) this version should be release in future.
Important thing is that Angular 2.0 will support only modern browsers for example IE10/11, because:
> Yes, there are still older browsers in use, but we’re targeting these newer models as they will be the primary ones in use by the time AngularJS 2 is ready and you good developers have had time to build apps on it. 
http://angularjs.blogspot.com/2014/03/angular-20.html

As you can see first version still maintaining and should be maintained few years, so migration to Angular 2.0 can be planned after Sylius 1.0.0 release.

**Comment by aramalipoor** [comment]
We've been using AngularJS about a year for our admin panel, we've faced a few issues which practically stopped us from moving faster. Now we're considering changing our tech-stack to a more robust and easier approach. Intuitive for customers and easy for developers with lowest technical debt as possible.

If  I could find some time I would be happy talk about the details, but these articles from Shopify team describes the issues very clearly. I can say I'm 90% agreed with things they saying about Admin panel!

* [Rebuilding the Shopify Admin: Improving Developer Productivity by Deleting 28,000 lines of JavaScript](http://www.shopify.com/technology/15646068-rebuilding-the-shopify-admin-improving-developer-productivity-by-deleting-28-000-lines-of-javascript/)
* [Y-Combinator topic](https://news.ycombinator.com/item?id=8523519)
* [History of Admin of Shopify](https://gist.github.com/kristianpd/f4c2e0aeb53d09f6def1)
* **[Summary of Their Findings](https://gist.github.com/kristianpd/2bc365a850dd0caa7293)**

Some libraries that can come of use:
* [Twine minimalistic two-way binding system](https://github.com/shopify/twine) by Shopify
* [Symfony TurbolinksBundle](https://github.com/helthe/TurbolinksBundle) port of the rails turbolinks gem

**Comment by kayue** [comment]
@aramalipoor Thanks for sharing. I am actually just started to build a backend with Angular 2! 

![warehouse](https://cloud.githubusercontent.com/assets/193112/12971985/2742affc-d0dc-11e5-8ca1-373455eba70b.png)

**Comment by aramalipoor** [comment]
@kayue This is good as an experiment, but IMO not for production! be careful of the forms/APIs/uploads/build processes/tests/TESTS/TEEESTS!! :D Protractor compared to Behat/Selenium is like comparing a bicycle to a submarine when you're talking about ocean! We couldn't access Symfony's container (e.g. for required fixtures), we couldn't smoothly talk to storefront, etc.

`ui-grid` is not ready for intuitive API integration, specially with Symfony, for example when Sylius have merged their GridBundle a Symfony/Twig-based admin panel is going to get all the benefits of that bundle but anyone using API is going to do another extra layer for the integration, specially for an MVC framework like AngularJS.

Maybe using Angular (2?) or React.JS only for rich-client parts seems like a better idea than having a full-stack SPA using a framework!

Anyway, wish you great success @kayue

**Comment by pjedrzejewski** [comment]
@kayue @aramalipoor Just as a side note, my grid implementation #3944 is format agnostic too, so you can get data in json/xml with all sorting/filtering functionality. That being said, rendering will be added in separate PR and will be done via Twig. But I can imagine having the rendering part in frontend and just consuming it via API calls. Grid fields configuration is in yml, so could be processed by this frontend framework, I guess.

**Comment by aramalipoor** [comment]
@pjedrzejewski Nice! This is a great news! Yes it could but the thing is it's better to keep main logic on the server since reading/manipulating/extending of those configuration is easier there. Imagine ImportExport bundle which trys to add a MassAction into every grid. Or a bundle trying to add a column based on some logic. Just like what OroPlatform is doing with their grids.

Moving this logic to the client-side will prevent flexibility and ease of development. IMO any application logic should stay on server and only components and widgets should be rendered on client. So even if you want to load grid data from an API lets not get the JSON and ask server to render that part in Twig (I know what you're thinking now :D I wouldn't say this a few months ago!!), and replace the table content! 

Anyway IMO keeping a single technology (Symfony) and use a very simple binding in JS seems like a good approach.

**Comment by pjedrzejewski** [comment]
Well for Sylius grids will be rendered via Symfony&Twig of course, but also accessible via API. I don't think we have time to implement anything frontend-side. We will have a look at this for 2.0 possibly, when Sylius API is more mature.

**Comment by aramalipoor** [comment]
Yeah, implementing a client-side solution specially for a complex thing like grid is more of a hassle than problem-solver. Wise choice :)

**Comment by pjedrzejewski** [comment]
I think we can close this, because we are sticking with jQuery + SemanticUI for 1.0 and will rely on Symfony for rendering. Of course you will be able to build your own shop/admin UI on top of API.

We will definitely consider providing decoupled UI on top of some framework for future versions.
